base/array.jl

```diff
@@ -545,7 +545,7 @@ else
 end
 
 _array_for(::Type{T}, itr, ::HasLength) where {T} = Vector{T}(undef, Int(length(itr)::Integer))
-_array_for(::Type{T}, itr, ::HasShape) where {T} = similar(Array{T}, axes(itr))::Array{T}
+_array_for(::Type{T}, itr, ::HasShape) where {T} = similar(Array{T}, axes(itr))
```

Comment:
This was required due to the change in `sort.jl` to a comprehension, which in turn was required because we were using the existence of a `Slice` to determine where a `:` was placed.  There could be another workaround for that if this is deemed undesirable… but it seems quite nice to be able to create similarly offset arrays simply with `[B[i,j] for (i,j) in axes(A)]`.

Reply:
This moves in the opposite direction of what I'm doing in #25861 (narrowing the type-assertion). However, if we have to pick one the two, I guess this one should have priority.

Reply:
Do we get the same benefit from `::AbstractArray{T,N}`?

Reply:
Also note that this change isn't completely predicated by the change in axis representation.  

While it's not been tested, the introduction of this assertion broke things like `[x for x in A]` if `A` is an offset array.  That had worked on 0.6.

Reply:
Looks like with the explicit introduction of `N`, we need no type assertion at all :tada: I've removed it from #25861, too.